draft-ietf-quic-transport.md

```diff
@@ -1069,6 +1069,18 @@ to cease using the connection IDs when requested can result in connection
 failures, as the issuing endpoint might be unable to continue using the
 connection IDs with the active connection.
 
+An endpoint MAY elect to only send or retransmit RETIRE_CONNECTION_ID frames
+with sequence numbers greater than or equal to the highest Retire Prior To field
+received minus its advertised active_connection_id_limit. This bounds the
```

Comment:
I'm afraid this advice might be inadequate.

It is possible to mount attacks without using Retire Prior To (see https://github.com/quicwg/base-drafts/issues/3509#issuecomment-597970501), and therefore the defense should rather not depend on the value of Retire Prior To.

I think it would be better to state that:
* An endpoint SHOULD limit the number of unacknowledged RETIRE_CONNECTION_ID frames (or the count of sequence numbers contained in those frames) that the endpoint tracks.
* That limitation SHOULD be no less than 2 * max_connection_id_limit.

I think that was what @martinthomson had in mind when he wrote https://github.com/quicwg/base-drafts/issues/3509#issuecomment-597425775.

`2 * max_connection_id_limit` is a better ballpark figure than `max_connection_id_limit`, because it gives some confidence to the issuer of CIDs that it can issue a new CID immediately when it receives RCID, without the fear of that causing some state to be lost in the peer.

Reply:
I guess we're going to have this discussion in two places, but I'm not sure the non-RPT case is a valid threat model.

I am a little concerned with your formulation that the endpoint could make bad choices about which RCIDs it decides to stick with. For example, we might never increase RPT and for some reason have a low sequence number that sticks around and is still being used. The server really better actually send the RCID.

I'm happy to haggle about the exact threshold, but my formulation frees the endpoint of all the state only when the client is doing something legal but silly.

Reply:
I didn't have a specific limit in mind.  I might consider active_connection_id_limit as the number (which bounds the total number of things you need to remember to 2 times that as you might also need to remember connection IDs for active use as well).  So I don't think that you need to specify a limit, just say that endpoints need to track the highest sequence number that needs to be retired and a contiguous block of sequences numbers for which it will attempt to send RETIRE_CONNECTION_ID frames.  That block might start at the bottom of the overall block of sequence numbers that require retirement.

Reply:
@martinthomson 
> I didn't have a specific limit in mind. I might consider active_connection_id_limit as the number (which bounds the total number of things you need to remember to 2 times that as you might also need to remember connection IDs for active use as well).

Thank you for the clarification. If that is the case, let my explain why I think `max_connection_id_limit * 2` is the correct value.

The draft currently allows an endpoint to request retirement of all the active CIDs that the peer retains at once. We explicitly define how the peer should behave in such case; see https://quicwg.org/base-drafts/draft-ietf-quic-transport.html#section-5.1.2-5. That means that the use of Retire Prior To might cause as much as *active_connection_id_limit* CIDs to be scheduled to be sent using RCID frames at once.

We also assume that the use of Retire Prior To is going to be infrequent.

Therefore, we need *active_connection_id_limit* in conjunction to the use of Retire Prior To.

Concurrently, the peer might be running business as usual, consuming CIDs as necessary. That means that we need some room for these too. We might argue that the active_connection_id_limit selected by the peer reflects the maximum pace that that peer is going to consume CIDs.

Based on this observation, my argument is that the sum of the two requirements makes sense; i.e. `max_connection_id_limit * 2`.

Reply:
You can only retire as many as you store, so the only way you could have more than max_connection_id_limit RETIRE_CONNECTION_ID frames outstanding is if you need to retransmit all of them due to loss and subsequently you've both received a new connection ID AND need to retire it.  Retiring it because you decided to use it typically takes at least a round trip, so it's most likely the subsequent retiring is due to the peer's Retire Prior To increasing again.

Given the above situation starts looking increasingly like an attack(or a bug) and sending RETIRE_CONNECTION_ID quickly for the Retire Prior To case isn't critical, I'd suggest max_connecion_id_limit is a better recommendation.

Reply:
How about we recommend limiting the number of RETIRE_CONNECTION_ID frames outstanding, but implementations SHOULD be willing to have at least max_connection_id_limit outstanding to allow a peer to retire all previously issued CIDs?

That isn't overly proscriptive, but gives a reasonable lower bound on how many implementations should allow.

Reply:
Consider the case where  client has sent an RCID frame. The server has received that frame and has sent an NCID frame carrying a replacement. But that server has not sent an ACK yet, or the ACK it has sent has been lost.

At this moment, the server might decide to update the CID encryption key, and try to retire all the CIDs that has been provided to the client. When the NCID frame carrying the request to retire reaches the client, the client is now required to track retirement of `max_connection_id_limit + 1` CIDs.

I agree that this is a corner case. But our previous consensus has been to allow an endpoint to request retirement of all the CIDs that have been provided at once. Assuming that we want to retain that consensus, requiring endpoints to track *more* than `max_connection_id_limit` is IMO the logical choice.

The effect of *not* requiring that would be that an endpoint that decides to retire the CIDs that it has provided to the peer has to do retirement in multiple phases; e.g., first request retirement of half of them, wait for RCID frames, then retire the other half. That's much more complicated than just saying recommending endpoints to track something like `max_connection_id_limit * 2`.

Reply:
I don't want to force retirement to be in multiple phases in edge cases like this, but I think it's fine to limit RCID to max_connection_id_limit outstanding frames.  Later, when some RCID frames get acknowledged, check if there are more CIDs that need retiring.  The case when the limit is exceeded needs to be handled in some way, so I'd rather have that code hit very infrequently than basically never.

An alternative I'd be ok with is to use max_connection_id_limit*2 and if the limit is exceeded, close the connection.  I don't see a problem with that, as we have limits like that already in our code, and it's very simple.  But it's fairly different from what we've been discussing.

Reply:
> I don't want to force retirement to be in multiple phases in edge cases like this, but I think it's fine to limit RCID to max_connection_id_limit outstanding frames. Later, when some RCID frames get acknowledged, check if there are more CIDs that need retiring. The case when the limit is exceeded needs to be handled in some way, so I'd rather have that code hit very infrequently than basically never.

I think I agree that tracking 2*max_connection_id_limit of RCID frames in just one way taking care of these edge cases. I'm fine with leaving it up to the implementations how they would deal with this problem.